gguf-py/gguf/gguf_reader.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import annotations
+
+import os
+from collections import OrderedDict
+from typing import NamedTuple, TypeVar
+
+import numpy as np
+import numpy.typing as npt
+
+if __name__ == "__main__":
+    import sys
+    from pathlib import Path
+
+    # Allow running file in package as a script.
+    sys.path.insert(0, str(Path(__file__).parent.parent))
+
+from gguf.constants import (
+    GGML_QUANT_SIZES,
+    GGUF_DEFAULT_ALIGNMENT,
+    GGUF_MAGIC,
+    GGUF_VERSION,
+    GGMLQuantizationType,
+    GGUFValueType
+)
+
+READER_SUPPORTED_VERSIONS = [2, GGUF_VERSION]
+
+
+class ReaderField(NamedTuple):
+    # Offset to start of this field.
+    offset: int
+
+    # Name of the field (not necessarily from file data).
+    name: str
+
+    # Data parts. Some types have multiple components, such as strings
+    # that consist of a length followed by the string data.
+    parts: [npt.NDArray] = []
```

Comment:
This is spelled `list[npt.NDArray]`. You may want to run mypy to find more errors like this.

Reply:
I thought I had it set up properly, but apparently not. These last set of changes clear up almost all the lints even running with `--strict`.

I really appreciate you taking the time to review the pull but please, _please_ don't just add commits to my pulls without talking to me about it first.

I guess I obliterated your changes, although I wasn't really trying to do that. I'm not really sure what the alternative is though since I had a big set of changes that now conflicted with my own pull.

Reply:
Noted. I personally prefer history-preserving operations like merges, it makes history easier to follow. Especially after a PR makes it out of draft status. It'll all get squashed in the end anyway.

I can save my cleanup for whenever you're ready. I'd rather not give you a pile of linter suggestions to correct if you don't have the same linter set up that I do.

Reply:
> I can save my cleanup for whenever you're ready. 

You can go ahead (and thanks) at least if you see this in the next few hours. I won't be working on it for a while, gotta sleep.

The main thing I'm concerned about is wanting to push something of my own and running into conflicts. I really don't enjoy dealing with merge conflicts. That's the main thing, but also it's not guaranteed I'll agree with whatever changes so I'd also like to have some idea of what's going to happen in advance.

> I personally prefer history-preserving operations like merges

I agree it's better. I've had some issues with other approaches in the past though, that's why I settled on my (less than idea) workflow of just squashing everything and rebasing. I would definitely prefer to keep the history though.